Included concepts and extensible points docs (#530)
OpenFreeEnergy/gufe: docs/concepts/included_models.rst

Alchemical methods performing free energy perturbation (FEP) between the two ``ChemicalSystem``\s of a ``Transformation`` will simulate these ``Component``\s using some sampling approach, obtaining enough information to derive a free energy difference estimate.

.. _transformation:

``Transformation``
------------------

A :class:`.Transformation` represents an alchemical transformation between two :ref:`ChemicalSystems <chemicalsystem>`.

``Transformation`` objects are often used as the edges of an :ref:`alchemicalnetwork`.
In addition to referencing the ``ChemicalSystem``\s it spans,
a ``Transformation`` also includes the :ref:`protocol` used to actually perform the alchemical transformation,
as well as an :ref:`componentmapping` specifying what portions of the :ref:`Components <component>` are being transformed across the ``ChemicalSystem``\s.

A ``Transformation`` functions as a container for all the information needed to obtain an estimate of the free energy difference between its two ``ChemicalSystem``\s.

.. _nontransformation:

``NonTransformation``
---------------------

> A :class:`.NonTransformation` represents non-alchemical sampling of a single :ref:`ChemicalSystem <chemicalsystem>`.

Author: Discussed in call; agreed that current wording sufficiently describes what a `NonTransformation` is, without going into detail as to why we designed it this way.

``Protocol`` objects are often used as part of a ``Transformation``, although they can be used on their own alongside ``ChemicalSystem``\s and ``ComponentMapping``\s (when needed) to obtain free energy difference estimates.
Individual ``Protocol`` subclasses obtain these estimates in a wide variety of ways, with varying domains of applicability and effectiveness.

The :meth:`.Protocol.create` method is used to generate :ref:`ProtocolDAGs <protocoldag>` that can be executed to produce :ref:`ProtocolDAGResults <protocoldagresult>`.
The :meth:`.Protocol.gather` method is used in turn to aggregate the contents of many :ref:`ProtocolDAGResults <protocoldagresult>` into a :ref:`ProtocolResult <protocolresult>`.

.. note::
    The :class:`.Protocol` is an *extensible point* of the library,
    and is intended to be subclassed to enable new applications.
    For details on how to create your own :class:`.Protocol` classes, see :ref:`howto-protocol`.

.. _protocoldag:

``ProtocolDAG``
^^^^^^^^^^^^^^^

> A :class:`.ProtocolDAG` is an executable artifact that performs a :ref:`Protocol <protocol>`.

Author: Switched to "object" for clarity.

Data models included in **gufe**
================================

The core of the **gufe** data model is the :class:`.GufeTokenizable` class,
but **gufe** features more than just this base data structure.

In order to ensure interoperability,
**gufe** also defines classes of objects that represent the core chemistry and alchemistry of a free energy pipeline,
including molecules, chemical systems, and alchemical transformations.

> This provides a shared language that tools in the ecosystem use.

Reviewer: ```suggestion
This provides a shared language that is used by tools across the OpenFE ecosystem.
```
Maybe this is more precise?

Generally speaking, :ref:`ChemicalSystems <chemicalsystem>` can be thought of as the *what* or the *nouns* that we are simulating,
:ref:`Transformations <transformation>` are the *how* or the *verbs* that encode how we are simulating these objects and moving between them,
and an :ref:`alchemicalnetwork` groups all of these together.

.. image:: ../_static/alchemical_network_diagram.svg
    :alt: The ``GufeTokenizable`` representation of an ``AlchemicalNetwork``.

.. note::

    Some of these classes are designed to be subclassed, and constitute the *extensible points* of the library.
    These include the following; see the **How-To Guide** for more information on how to extend from each:

    1. :ref:`Component <component>` : :ref:`howto-component`
    2. :ref:`Protocol <protocol>` : :ref:`howto-protocol`
    3. :ref:`ComponentMapping <componentmapping>` : :ref:`howto-componentmapping`
    4. :ref:`AtomMapping <atommapping>` : :ref:`howto-atommapping`
    5. :ref:`AtomMapper <atommapper>` : :ref:`howto-atommapper`

.. _component:

``Component``
-------------

The :class:`.Component` class represents a portion of a system of molecules,

> with a single ``Component`` capable of representing anything from an individual drug-like molecule, an entire protein, or (even the concept of) a solvent with ions.

Author: Works for me! Thank you!

> These are often used to define the *components* of a :ref:`chemicalsystem`, which form the nodes of an :ref:`alchemicalnetwork`.

Reviewer: ```suggestion
``Components`` are often used as the building blocks of :ref:`chemicalsystem`, which form the nodes of an :ref:`alchemicalnetwork`.
```
Saying that a component is a component isn't super clear.

The same ``Component`` may be present within multiple ``ChemicalSystem``\s, such as a :class:`.ProteinComponent` in an ``AlchemicalNetwork`` featuring relative binding transformations between ligands.

As another distinct example: the :class:`.SmallMoleculeComponent` class is used to form the nodes of a :ref:`ligandnetwork`.
This is useful for representing relative transformations between a series of small molecules without invoking the additional complexity of an :ref:`alchemicalnetwork`.

.. note::
    The :class:`.Component` is an *extensible point* of the library,
    and is intended to be subclassed to enable new applications.
    For details on how to create your own :class:`.Component` classes, see :ref:`howto-component`.

.. _chemicalsystem:

``ChemicalSystem``
------------------

A :class:`.ChemicalSystem` represents a complete system of molecules,
and is often composed of multiple :ref:`Components <component>`.

These are most often used as nodes of an :ref:`alchemicalnetwork`, with pairs of ``ChemicalSystem``\s connected by :ref:`Transformations <transformation>`.

> Because a ``ChemicalSystem`` functions as a kind of container of :ref:`Components <component>`, more than one ``ChemicalSystem`` can feature the same ``Component``\s.

Reviewer: ```suggestion
Because a ``ChemicalSystem`` functions as a kind of container of references to :ref:`Components <component>`, more than one ``ChemicalSystem`` can feature the same ``Component``\s.
```

This allows even very large ``AlchemicalNetwork``\s to be relatively small in memory, as only a few large ``Component``\s like :class:`.ProteinComponent`\s may be shared among hundreds of ``ChemicalSystem``\s.

When used as inputs to a ``Transformation``, ``ChemicalSystem``\s represent the set of ``Component``\s for which a free energy difference will be estimated.
Alchemical methods performing free energy perturbation (FEP) between the two ``ChemicalSystem``\s of a ``Transformation`` will simulate these ``Component``\s using some sampling approach, obtaining enough information to derive a free energy difference estimate.

.. _transformation:

``Transformation``
------------------

A :class:`.Transformation` represents an alchemical transformation between two :ref:`ChemicalSystems <chemicalsystem>`.

``Transformation`` objects are often used as the edges of an :ref:`alchemicalnetwork`.

> In addition to referencing the ``ChemicalSystem``\s it spans,

Reviewer: ```suggestion
In addition to referencing the two ``ChemicalSystem``\s it spans,
```

a ``Transformation`` also includes the :ref:`protocol` used to actually perform the alchemical transformation,
as well as an :ref:`componentmapping` specifying what portions of the :ref:`Components <component>` are being transformed across the ``ChemicalSystem``\s.

A ``Transformation`` functions as a container for all the information needed to obtain an estimate of the free energy difference between its two ``ChemicalSystem``\s.

.. _nontransformation:

``NonTransformation``
---------------------

A :class:`.NonTransformation` represents non-alchemical sampling of a single :ref:`ChemicalSystem <chemicalsystem>`.

In the context of an :ref:`alchemicalnetwork`, a ``NonTransformation`` is effectively a self-loop, featuring the same ``ChemicalSystem`` at either end.

> Similar to a :ref:`Transformation <transformation>`, it features a :ref:`protocol` used to perform sampling on its ``ChemicalSystem``, but does not feature a :ref:`componentmapping` since none is required for this.

Reviewer: ```suggestion
Similar to a :ref:`Transformation <transformation>`, it features a :ref:`protocol` used to perform sampling on its ``ChemicalSystem``, but does not feature a :ref:`componentmapping`, since there is no second ``ChemicalSystem``.
```
it's not just not required, it cannot/should not be defined.
